### ISSUE TITLE
Morality Metrics

### ISSUE THREAD

**Issue by meatballs** [Axelrod-Python/Axelrod#220]
Add morality metrics to the resultset as described in http://www.scottaaronson.com/morality.pdf

**Comment by meatballs** [comment]
Code for the work described in the paper is here: https://github.com/tscizzle/IPD_Morality

**Comment by marcharper** [comment]
@tscizzle Do you have any interest in adding this functionality to our library?

**Comment by meatballs** [comment]
I'll have a crack at this. It looks interesting!

I'd appreciate any thoughts on how to return the results.....

There seems to be 4 metrics: Co-Operation, Good Partner, EigenJesus and EigenMoses. In each case, for a given tournament, each player receives a single numeric rating.

That suggests a list for each metric, but I think that then gives us two options:

1. Return each list separately as a property of ResultSet (e.g. ResultSet.EigenJesus)
1. Return a dictionary (e.g. ResultSet.morality_metrics) with the metric name as the key and the list as the value

I quite like option 2, but it's fairly marginal and there may be other ways.

**Comment by marcharper** [comment]
I usually prefer dictionaries unless there is a good reason for something else, simply because it's easy to figure out what a dictionary holds using .keys() or .items(). Plus it's usually easy to coax the data into another format or object. But if you want to do something that ResultsSet makes easier with the data right away it may make since to go with 1.

Are you planning to enumerate the strategies to make the list? I often use dictionaries for that purpose as well (you could use the names of the strategies as keys) so that I don't have to worry about how the data is enumerated, and since enumerating hashables later is easy. But I guess that depends on how the calculation is being done -- if you are using say numpy to compute the eigenvector than I guess you have to enumerate the players anyway.

**Comment by marcharper** [comment]
Looks like @tscizzle used the "take large powers" approach using a numpy matrix to compute the principle eigenvector. I used a similar approach in my stationary distribution library (but for a sparse matrix). FYI I found that to give better results than numpy's eigenvector computation functions.

**Comment by drvinceknight** [comment]
If we're doing eigen analysis perhaps going to numpy makes sense? We had long discussions about _to numpy or not to numpy... (#158) I was the main person opposed to using numpy as the numerics we're doing are not heavy and I wanted the library to be as close to base python as possible (it can run without matplotlib for example) but I think I was perhaps wrong (or at least certainly am when we're going to be computing eigen*).

Given that we could potentially have quite a large matrix (theoretically of any size eventually once all the strategies come flowing in :)) I think the 'large powers' approach make sense. (Although I'm impressed/surprised that it gave better results than numpy's native tools).

If we are going to use numpy for this (which I again think makes complete sense) it would probably make sense to start porting other things to use numpy... 

Also for what it's worth installing numpy on travis needs a bit of tlc. Here's the travis.yml file I'm using for another project that needs numpy: https://github.com/drvinceknight/TwoThirds/blob/master/.travis.yml

Finally: :+1: for dictionaries.

**Comment by drvinceknight** [comment]
> if you are using say numpy to compute the eigenvector than I guess you have to enumerate the players anyway.

If you build the dictionary of states _first_ then you can use that arbitrary ordering of keys to build the numpy matrix (that's how I've done things in the past).

**Comment by meatballs** [comment]
Given that the computation of the eigenvectors is likely to be significant, I suggest we put a flag on a tournament to indicate whether it should be performed.

We'll need to produce a co-operation matrix and that relies upon the individual interactions within a round robin. I suggest we always calculate and return that matrix in the resultset, but only use it to compute the metrics themselves if requested to do so.

**Comment by drvinceknight** [comment]
Yeah that sounds good to me. Certainly shouldn't be calculating eigen stuff if we don't want to.

**Comment by meatballs** [comment]
Footnote 11 in the morality paper mentions the fact that the principal eigenvalue of the co-operation matrix can be used a measure of the total 'goodness' of the environment. 

I think we should calculate and return that too.

@uglyfruitcake - that might provide some evidence to back your opinion that the strategies in the library so far are excessively 'nice.'

**Comment by drvinceknight** [comment]
> I think we should calculate and return that too.

Why not.

> 
> @uglyfruitcake - that might provide some evidence to back your opinion that the strategies in the library so far are excessively 'nice.'

Most are probably nice as that was one of the findings that Axelrod presented, in that they need to be nice to do well. That's what's cool about the backstaber strategy (which I'm looking forward to seeing :)) as well as the Meta Hunter :)

**Comment by meatballs** [comment]
For the computation of the Co-Operation matrix, we should also use some form of deterministic cache. I think there are two options:

1. Create a new cache using similar ideas to the existing scores cache
1. Replace the scores cache with a cache of the interactions between deterministic players. This could be used to calculate both the scores and the co-operation.

Option 2 sounds the cleaner, DRYer option but it's more work!

**Comment by meatballs** [comment]
More complications....

At the moment the RoundRobin.play() method computes and returns the payoff matrix. We now need to add the cooperation matrix. Two ways of doing this:

1. The method return both matrices
1. Both matrices become attributes of the class which are populated by the play method but accessed independently.

In either case, we'll have to change every call to the play method to take account of the change. I think option 2 is more future proof.

**Comment by drvinceknight** [comment]
> To start, I could get things up and running without any caching at all, but we'll need to get this right at some point.

This sounds sensible. Once tests are in place we/you can refactor etc... I agree that 2 sounds nicer.

> At the moment the RoundRobin.play() method computes and returns the payoff matrix. We now need to add the cooperation matrix. Two ways of doing this:

I'm relatively indifferent but agree that 2 is nicer. Less confusing to a user who is using the library alone.

**Comment by langner** [comment]
There is another alternative for a quick hack: increase the dimensionality of the returned matrix. But I agree that 2. is nicer.

**Comment by meatballs** [comment]
So, something approximating a plan for this one looks like:

1. Refactor the payoffs matrix to become an attribute of the RoundRobin class
1. Add the cooperation matrix as an attribute of the RoundRobin class
1. Add a morality dictionary to the ResultSet class
1. Add a flag to the Tournament class to control whether the morality dictionary is populated
1. Create a morality calculator class with methods to compute the various morality metrics

**Comment by drvinceknight** [comment]
Sounds good to me.

**Comment by meatballs** [comment]
If anyone else fancies joining in, I can stub out the morality calculator class and leave its implementation and test suite as a separate exercise. Don't worry if not - I'm happy to do the lot.

**Comment by marcharper** [comment]
I can write the eigenvector calculations and tests at least.

**Comment by tscizzle** [comment]
Hey guys, I would offer to help with code on this part but at this moment in time I wouldn't be able to be reliable so I think I'd just be a liability. But if there are any questions I can help with I'd be more than happy to discuss!

**Comment by marcharper** [comment]
Hi @tscizzle! I'm going to work on adding in the rest of your bots so we can reproduce your calculations. Do you mind telling me what your source was? Most look like they are from Axelrod's original tournament.

**Comment by drvinceknight** [comment]
@marcharper if I can help with checking if some strategies are implemented already or not let me know :)

**Comment by tscizzle** [comment]
They come from a few places:
- Axelrod's "Evolution Of Cooperation" describes a bunch of bots, some from the first tournament and some from the second, and I believe even some just to illustrate concepts but didn't show up in either tournament.
- I got a few from http://www.prisoners-dilemma.com/ under "Strategies for IPD".

I can also explain any given bot if it's not clear (they should be doc'd but definitely I could have missed one or left out details.)

**Comment by meatballs** [comment]
> I can write the eigenvector calculations and tests at least.

@marcharper That would be great. Thank you

**Comment by marcharper** [comment]
@tscizzle, I can look at your code if that's ok with you -- which ones are your originals? I want to give you proper credit. (Also, if you don't mind adding a license to your repository, that would appreciated.)

@drvinceknight Does anyone have a copy of Axelrod's book? If not there are plenty of cheap copies on Amazon and I'll pick one up.

**Comment by drvinceknight** [comment]
> @drvinceknight Does anyone have a copy of Axelrod's book? If not there are plenty of cheap copies on Amazon and I'll pick one up.

I was just thinking that (I've got a copy and wasn't sure of easiest way to send over or whatever was best). Just emailed you Axelrod's two papers (wasn't sure if you were paywalled out...) - I did not think the book gave a tonne more explanation than is in the papers (again for the second tournament, a lot of strategies are not fully explained).

**Comment by drvinceknight** [comment]
But yeah, I've got a copy, I can always snail mail it to you although I doubt that would be much cheaper than amazon, another option would be for us to work through together but my time is pretty eaten up this Summer so I might slow you down...

**Comment by marcharper** [comment]
@drvinceknight I'll get a copy, it's nbd. I just need something that spells out the more complex strategies in detail. Thanks for the offer!

**Comment by drvinceknight** [comment]
> more complex strategies in detail

Unless this was in the chapters I skimmed I'm not sure you'll find that in that book, am I wrong though @tscizzle ? Still definitely worth getting though :)

**Comment by tscizzle** [comment]
Absolutely still worth getting. One of the best books ever. But you're right, the book only includes explanations of a handful of the strategies.

**Comment by tscizzle** [comment]
I will go through the bots in my repo as well as add a license soon.

**Comment by tscizzle** [comment]
Ok, it turns out all those bots do come from those 2 sources, I did not make any up that are in my repo. So anything in my repo that isn't in this list (http://www.prisoners-dilemma.com/strategies.html) is from "Evolution Of Cooperation".

**Comment by tscizzle** [comment]
License added to IPD Morality. That work?

**Comment by marcharper** [comment]
@tscizzle Thanks so much!

@meatballs I pushed a branch for the power iteration calculation of the principle eigenvalue on my repository -- see eigen.py and test_eigen.py (wasn't sure where to put them yet)

**Comment by meatballs** [comment]
@marcharper I've merged those changes into branch 220 which is now available on the main repo and I'll be working on this ticket on that branch (when I finally get the time)!

**Comment by meatballs** [comment]
Progress.....

* This isn't going to work fully without the caching in place. We use the cache to bypass the play altogether - and that means we don't get the cooperation rates if the players already exist in the scores cache. At the moment, I'm populating the cooperation matrix whenever an interaction is played, but setting it to zero if not.

* I've set both the cooperation and payoff matrices as attributes of RoundRobin, but I've left the play() method returning the payoffs so as to reduce the changes made elsewhere

* I've refactored some of the stuff in the play method out into separate methods (but there's more here to be done). That increases the unit tests required but I'd rather pick that up and finish the job properly as part of #191
